projects/stone_cx/robot.cc

```diff
@@ -1,170 +1,413 @@
+    virtual std::array<float, 2> getSpeed() override
+    {
+        // **TODO** sideways awesome
+        const float speed =  m_Omni.getForwards() * m_VelocityScale;
+        return {speed, speed};
+    }
 
-            // Reduce filtered flow - summing along rows
-            cv::reduce(flowXSum, flowSum, 1, cv::REDUCE_SUM);
+private:
+    //------------------------------------------------------------------------
+    // Members
+    //------------------------------------------------------------------------
+    const Robots::Omni2D &m_Omni;
+    const std::array<radian_t, 2> m_PreferredAngleTN2;
+    const float m_VelocityScale;
+};
 
-            // Filter speed and set atomic
-            prevSpeed = (alpha * flowSum.at<float>(0, 0)) + ((1.0f - alpha) * prevSpeed);
-            speed = prevSpeed;
+//---------------------------------------------------------------------------
+// SpeedSourceVicon
+//---------------------------------------------------------------------------
+class SpeedSourceVicon : public SpeedSource
+{
+public:
+    SpeedSourceVicon(const Vicon::UDPClient<Vicon::ObjectDataVelocity> &vicon,
+                     const std::array<radian_t, 2> &preferredAngleTN2 = {45_deg, -45_deg},
+                     float speedScale = 5.0f)
+    :   m_Vicon(vicon), m_PreferredAngleTN2(preferredAngleTN2), m_SpeedScale(speedScale)
+    {
+    }
+
+    //------------------------------------------------------------------------
+    // SpeedSource virtuals
+    //------------------------------------------------------------------------
+    virtual std::array<float, 2> getSpeed() override
+    {
+        // Read data from VICON system
+        auto objectData = m_Vicon.getObjectData();
+        const auto &velocity = objectData.getVelocity();
+        const auto &attitude = objectData.getPose().attitude();
+
+        const radian_t heading = -attitude[2];
+
+        std::array<float, 2> speedTN2;
+        for(size_t j = 0; j < Parameters::numTN2; j++) {
+            speedTN2[j] = (sin(heading + m_PreferredAngleTN2[j]) * m_SpeedScale * velocity[0].value()) +
+                          (cos(heading + m_PreferredAngleTN2[j]) * m_SpeedScale * velocity[1].value());
         }
+        return speedTN2;
     }
 
+private:
+    const Vicon::UDPClient<Vicon::ObjectDataVelocity> &m_Vicon;
+    const std::array<radian_t, 2> m_PreferredAngleTN2;
+    const float m_SpeedScale;
+};
+
+//---------------------------------------------------------------------------
+// MotorControllerOmni2D
+//---------------------------------------------------------------------------
+class MotorControllerTank : public MotorController
+{
+public:
+    MotorControllerTank(Robots::Tank &tank)
+    :   m_Tank(tank)
+    {
+    }
+    
+    virtual void drive(float left, float right, bool log) override
+    {
+        // Calculate differential steering signal
+        const float steering = left - right;
+        if(log) {
+            LOGI << "Steer:" << steering;
+        }
```

Comment:
It might be easier just to use ``LOGD`` and then if you do want to see the output you can just run ``LOG_LEVEL=DEBUG ./my_program``

Reply:
ahh - is **that** how you set the bob robotics log-level :smile: